File: app/src/main/java/com/coppermobile/mysamplemvvmdatabindinglivedata/ViewModelFactory.java
package com.coppermobile.mysamplemvvmdatabindinglivedata;

import android.app.Application;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.coppermobile.mysamplemvvmdatabindinglivedata.utils.Injection;
import com.coppermobile.mysamplemvvmdatabindinglivedata.viewmodels.CommentsViewModel;
import com.coppermobile.mysamplemvvmdatabindinglivedata.viewmodels.DishViewModel;

public class ViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    private static ViewModelFactory INSTANCE;
    private final Application mApplication;

    public static ViewModelFactory getInstance(Application application) {

        if (INSTANCE == null) {
            synchronized (ViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new ViewModelFactory(application);
                }
            }
        }
        return INSTANCE;
    }

    private ViewModelFactory(Application application) {
        mApplication = application;
    }

    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(DishViewModel.class)) {
            //noinspection unchecked
            return (T) new DishViewModel(mApplication, Injection.provideDishRepository(mApplication));
        } else if (modelClass.isAssignableFrom(CommentsViewModel.class)) {
            //noinspection unchecked
            return (T) new CommentsViewModel(mApplication,Injection.provideCommentsRepository(mApplication));
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: app/src/main/java/com/coppermobile/kotlinsamplemvvmlivedataroom/ViewModelFactory.kt
package com.coppermobile.kotlinsamplemvvmlivedataroom

import android.app.Application
import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import com.coppermobile.kotlinsamplemvvmlivedataroom.utils.Injection
import com.coppermobile.kotlinsamplemvvmlivedataroom.viewmodels.CommentsViewModel
import com.coppermobile.kotlinsamplemvvmlivedataroom.viewmodels.DishViewModel

class ViewModelFactory private constructor(private val application: Application) : ViewModelProvider.NewInstanceFactory() {

    companion object {

        @Volatile
        private var INSTANCE: ViewModelFactory? = null

        fun getInstance(application: Application): ViewModelFactory {
            if (INSTANCE == null) {
                synchronized(this) {
                    INSTANCE = ViewModelFactory(application)
                }
            }
            return INSTANCE!!
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(DishViewModel::class.java)) {
            return DishViewModel(application, Injection.provideDishRepository(application)) as T
        } else if (modelClass.isAssignableFrom(CommentsViewModel::class.java)) {
            return CommentsViewModel(application, Injection.provideCommentsRepository(application)) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class: " + modelClass.name)
    }
}